flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueueImpl.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.operators.lifecycle.event;
+
+import org.apache.flink.util.function.RunnableWithException;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.LinkedBlockingQueue;
+
+class TestEventQueueImpl implements TestEventQueue {
```

Comment:
I think this is somewhat too complicated and hard to follow with those handlers, listeners and runnables. It took me more then 10 minutes to understand this code (and I had to open it in the IDE to click through all of the usages.

Do we need all of those extra layers of abstractions and indirections? As of now this simplified proposal https://github.com/pnowojski/flink/tree/f21090 would work just as fine. (note my other comment about squashing TestEventQueue interface with it's implementations).

Reply:
As commented [above](https://github.com/apache/flink/pull/16773#discussion_r688929533)
> I totally agree that the internals of TestEventQueueImpl should be simplified (those parts are some leftovers after changing the logic and refactoring , sorry about that). I added 77ec664 to address that, PTAL.

Reply:
Unfortunately, I had to revert some simplifications to address a race condition in 7913299066a7fdd2219634f986445d4bf4d395a1.
In particular, I added `addListener` to the EventQueue, so that it can be added before executing a command.